packages/flutter/lib/src/painting/linear_border.dart

```diff
@@ -0,0 +1,376 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:ui' show lerpDouble;
+
+import 'package:flutter/foundation.dart';
+
+import 'basic_types.dart';
+import 'borders.dart';
+import 'edge_insets.dart';
+
+/// Defines the relative size and alignment of one <LinearBorder> edge.
+///
+/// A [LinearBorder] defines a box outline as zero to four edges, each
+/// of which is rendered as a single line. The width and color of the
+/// lines is defined by [LinearBorder.side].
+///
+/// Each line's length is defined by [size], a value between 0 and 1
+/// (the default) which defines the length as a percentage of the
+/// length of a box edge.
+///
+/// When [size] is less than 1, the line is aligned within the
+/// available space according to [alignment], a value between -1 and
+/// 1.  The default is 0, which means centered, -1 means align on the
+/// "start" side, and 1 means align on the "end" side. The meaning of
+/// start and end depend on the current [TextDirection], see
+/// [Directionality].
+@immutable
+class LinearBorderEdge {
+  /// Defines one side of a [LinearBorder].
+  ///
+  /// The values of [size] and [alignment] must be between
+  /// 0 and 1, and -1 and 1 respectively.
+  const LinearBorderEdge({
+    this.size = 1,
+    this.alignment = 0,
+  }) : assert(size >= 0 && size <= 1);
+
+  /// A value between 0 and 1 that defines the length of the edge as a
+  /// percentage of the length of the corresponding box
+  /// edge. Default is 1.
+  final double size;
+
+
+  /// A value between -1 and 1 that defines how edges for which [size]
+  /// is less than 1 are aligned relative to the corresponding box edge.
+  ///
+  ///  * -1, aligned in the "start" direction. That's left
+  ///    for [TextDirection.ltr] and right for [TextDirection.rtl].
+  ///  * 0, centered.
+  ///  * 1, aligned in the "end" direction. That's right
+  ///    for [TextDirection.ltr] and left for [TextDirection.rtl].
+  final double alignment;
```

Comment:
It seems to me like we could use an `AlignmentGeometry` here, using the cross axis part of the alignment to indicate the equivalent of `BorderSide.strokeAlign`. People are familiar with `Alignment`, and it already has support for directionality in the form of `AlignmentDirectional` and it supports lerping. Plus it has the right name for this property. :-)

Reply:
I hadn't thought of enabling developers to also specify the border's "stroke" alignment. This will complicate the border's layout considerably because the implementation will need to decide if adjacent edges overlap after applying the cross-axis (stroke) alignment.

Reply:
You could maybe use the stroke alignments of the adjacent sides to modify what "full length" means, then a size of 1.0 would still not overlap, it would be just the right length to fit between the adjacent sides.

Reply:
Right, that's what happens (OK _should happen_) now. If you specify top, bottom, and start, the start edge should be shortened so that it doesn't overlap the bottom and top edges.  And the overlap only occurs if the horizontal edges size and alignment permits it. So, to be fair, this doesn't really complicate the implementation, it just adds some additional cases to test.

Testing has been an obstacle. If LinearBorder draws 2 or more Paths (one per edge), it's not clear how to check them. There's no way to say "one of the paths is expected to be ..." or "the 2nd path is expected to be...".

Reply:
I tried this but found that mapping from the existing Alignment constants to 2D LinearBorder edge alignments to be rather confusing. Will leave things as they are and consider adding a crossAxisAlignment property to LinearBorderEdge in the future, if there's demand for it.